### PR TITLE
[TorchTidy] Add Pattern to detect Synchronous Data Loader

### PR DESCRIPTION
**Pull request by davidchencsl**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82261
* #82248
* #81941
* #81921
* __->__ #81740

Summary: By setting num_workers > 0 in DataLoader, we can achieve async data loading, which is non blocking to the computation. This helps speed up the training process. By matching the call structure, we can detect if we are using Synchronous Data Loader.

Test Plan:
Added test in test.profiler.py

Differential Revision: [D38082644](https://our.internmc.facebook.com/intern/diff/D38082644)

### CODE REVIEW

File: torch/profiler/_pattern_matcher.py [hunk 1]
```diff
@@ -263,7 +264,7 @@ def skip(self):
         # Anything less than sm_80 is not Ampere which doesn't support TF32
         has_tf32 = all(
             int(arch[3:]) >= 80 for arch in torch.cuda.get_arch_list())
-        return has_tf32 is False
+        return has_tf32 is False or super().skip or not self.prof.record_shapes
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:267]: Looks like there are several updates to prior PRs that you discovered here. You should move them to their respective PRs.

File: torch/profiler/_pattern_matcher.py [hunk 2]
```diff
@@ -345,6 +345,58 @@ def match(self, event: _ProfilerEvent):
...
+    def __init__(self, prof: profile):
+        super().__init__(prof)
+        self.name = "Synchronized DataLoader Pattern"
+        self.description = (
+            "Detected DataLoader running with synchronized implementation. "
+            "Please enable asynchronous dataloading by setting num_workers > 0 when initializing DataLoader."
+        )
+        # We precompile the regex to avoid recompiling it every time we match
+        # The four back slashes are to escape windows path separator
+        seperator = "\\\\" if os.sep == "\\" else os.sep
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:375]: If you steal the binding from `https://github.com/pytorch/pytorch/pull/81895` then you won't have to mess with regex; you'll be able to directly access the underlying fields. (And you can just compare the filename with `os.path.join(["torch", "utils", "data", "dataloader.py])` to get the right path format.

File: torch/profiler/_pattern_matcher.py [hunk 3]
```diff
@@ -345,6 +345,58 @@ def match(self, event: _ProfilerEvent):
...
+            r"torch/utils/data/dataloader.py\([0-9]+\): __iter__$"
+            .replace("/", seperator))
+        self.get_iterator_regex = re.compile(
+            r"torch/utils/data/dataloader.py\([0-9]+\): _get_iterator$"
+            .replace("/", seperator))
+        self.check_worker_regex = re.compile(
+            r"torch/utils/data/dataloader.py\([0-9]+\): check_worker_number_rationality$"
+            .replace("/", seperator))
+
+    def match(self, event: _ProfilerEvent):
```
Comment (robieta) [inline torch/profiler/_pattern_matcher.py:386]: This could use some comments to explain your reasoning. I think it also needs some sort of check for "waiting". If my dataloader is able to keep up with training (maybe I already have the data in memory, maybe I'm handling parallelism elsewhere, maybe it's just really cheap) then `num_workers=0` is appropriate.

Comment (robieta) [reply torch/profiler/_pattern_matcher.py:386]: (You'll have to override this for test stability, of course.)

**Review by robieta — CHANGES_REQUESTED** [review]

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

File: torch/profiler/_pattern_matcher.py (hunk 3 above)
Comment (davidchencsl) [reply torch/profiler/_pattern_matcher.py:395]: I think I should add a TODO here. I was planning to add a check to the idle time during data loading, but I want to do it after kineto type is in.

File: torch/profiler/_pattern_matcher.py (hunk 2 above)
Comment (davidchencsl) [reply torch/profiler/_pattern_matcher.py:384]: One advantage for me to use regex here is that I can precompile the pattern as regex object. I don't need to process all the event name. Versus applying special treatment for all the event name

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

**Comment by davidchencsl** [comment]
@davidchencsl has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38082644).

**Review by robieta — APPROVED** [review]
LGTM
